From 1476e8109a152b07006479c0ab8d0ee22fe59c95 Mon Sep 17 00:00:00 2001
From: Weian Chen <weian.chen@intel.com>
Date: Mon, 30 Jul 2012 04:40:25 -0400
Subject: [PORT FROM MAIN] Awesomeplayer: Add SEEK_COMPETE event notification to listener

BZ: 48604

Original patch:
R3: 30123(BZ: 18523)
R2: 19887/22919(BZ: 7394 13217)

For audio + video.mp4 clip, when event PLAYBACK_COMPLETE occurs at the same time for
both audio and video, seek operation is ignored. However AwesomePlayer's listener doesn't
update on time with correct state. So Add SEEK_COMPLETE notification to listener when
PLAYBACK_COMPLETE occurs in AwesomePlayer.

Change-Id: I0e20cb45d496511bfb226a84962d75da6367d134
Orig-Change-Id: I69b22fd41d0d44dc8afc740d55433baa881a77bf
Signed-off-by: Weian Chen <weian.chen@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
Signed-off-by: Tong, Bo <box.tong@intel.com>
Reviewed-on: http://android.intel.com:8080/74922
Reviewed-by: lab_cactus <lab_cactus@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 media/libstagefright/AwesomePlayer.cpp |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/AwesomePlayer.cpp b/media/libstagefright/AwesomePlayer.cpp
index bd28118..5bbd305 100644
--- a/media/libstagefright/AwesomePlayer.cpp
+++ b/media/libstagefright/AwesomePlayer.cpp
@@ -842,6 +842,13 @@ void AwesomePlayer::onStreamDone() {
 
         pause_l(true /* at eos */);
 
+        // If audio hasn't completed MEDIA_SEEK_COMPLETE when play back complete echo,
+        // notify MEDIA_SEEK_COMPLETE to observer immediately for state persistance.
+        if (mWatchForAudioSeekComplete) {
+            notifyListener_l(MEDIA_SEEK_COMPLETE);
+            mWatchForAudioSeekComplete = false;
+        }
+
         modifyFlags(AT_EOS, SET);
     }
 }
@@ -1291,6 +1298,11 @@ status_t AwesomePlayer::getPosition(int64_t *positionUs) {
         *positionUs = 0;
     }
 
+    // set current position to duration when EOS.
+    if (mFlags & AT_EOS) {
+        *positionUs = mDurationUs;
+    }
+
     return OK;
 }
 
-- 
1.7.4.1

